lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -92,3 +105,377 @@ def gen_month_range(prefix, index, start, end):
+        """
+        monthResults = list()
+        queryString = ""
+        first_month = start.replace(day=1)
+        last_month = end + relativedelta(day=31)
+        for m in rrule.rrule(rrule.MONTHLY, dtstart=first_month, until=last_month):
+            monthResults.append(m.strftime("%Y-%m"))
+
+        # TODO: hardcoding the index here is risky. We need a framework to
+        # help the web services understand index versions and template
+        # formats, probably by building a persistent database from the
+        # index template documents at startup. This is TBD.
+        for monthValue in monthResults:
+            if index == "v4.result-data.":
+                queryString += f"{self.prefix + index + monthValue}-*,"
+            else:
+                queryString += f"{self.prefix + index + monthValue},"
+        return queryString
+
+    def assemble(
+        self, json_data: Dict[AnyStr, Any], string_data: Dict[AnyStr, AnyStr]
+    ) -> Dict[AnyStr, Any]:
+        """
+        assemble Assemble the Elasticsearch parameters
+
+        This must be overridden by the subclasses!
+
+        Args:
+            json_data: Input JSON payload, processed by type conversion
+
+        Returns:
+            Request: Package the path, json, and parameters to pass
+                    to Elasticsearch
+                path: The path part of the Elasticsearch URI
+                json: The JSON query to pass to Elasticsearch
+                params: Query parameters to pass to Elasticsearch
+        """
+        pass
+
+    def postprocess(self, response: Response) -> Dict[AnyStr, Any]:
+        """
+        postprocess Given the Elasticsearch Response object, construct a JSON
+                dict to be returned to the original caller.
+
+        This must be overridden by the subclasses!
+
+        Args:
+            response: Response object
+
+        Returns:
+            The JSON payload to be returned to the caller
+        """
+        pass
+
+    def _call(
+        self,
+        method: Callable,
+        json_data: Dict[AnyStr, Any],
+        string_data: Dict[AnyStr, AnyStr],
+    ):
+        """
+        _call Perform the requested call to Elasticsearch, and handle any
+        exceptions.
+
+        Args:
+            method: Any requests HTTP method (e.g., requests.post)
+            json_data: Type normalized client JSON input
+            string_data: Original client JSON input
+
+        Returns:
+            Postprocessed JSON body to return to client
+        """
+        try:
+            es_request = self.assemble(json_data, string_data)
+
+            # The assembler returns the uri "path" as a dict value for
+            # consistency. However requests.post() defines "url" as a
+            # positional parameter, and does not allow unexpected kwargs,
+            # so we need to remove this key.
+            path = es_request.get("path")
+            del es_request["path"]
```

Comment:
I haven't experimented with exactly how the `requests` methods handle their kwargs. Is it OK to specify `headers=None` for example? It would be safer to build a new dict with only the keys actually returned from the subclass, and I didn't feel like doing that. Just deleting the one I knew wasn't like the others was easier even if in some ways it's a little less clean. I'm not entirely happy with either strategy.

Reply:
I really don't like returning anonymous tuples. It is a bit better with type hinting because we can at least document the types inside; it's still harder to define exactly what they mean.

I'd thought of building a class, which is a cleaner way to do that, and then I tried to shift to the requests package `Request` class, but ran into other issues and dropped back to a simple `dict` which is essentially a named tuple. It's just convenience that by naming the keys properly I could reuse that as a requests kwarg ... except for that pesky url.

If I get pushed into changing this, I'll go for a new class, but I think that's overkill.

Reply:
OK, so you like the `dict` approach, I can live with that:  instead of a tuple, return a dictionary with two keys, `path` and `kwargs` (or something similar), and then
```
            es_response = method(
                return_dict['path'],
                **return_dict['kwargs']
            )
```
(obviously, you'll need to build a URL out of the path first, but you get the idea).

Sounds like a win.